tests/rpc/rpc_00_test.go

```diff
@@ -38,13 +39,14 @@ type testCase struct {
 	skip        bool
 }
 
-func getResponse(t *testing.T, test *testCase) interface{} {
+func getResponse(ctx context.Context, t *testing.T, test *testCase) interface{} {
```

Comment:
could this function not just accept a `time.Duration` for timeout and handle context creation and cancellation within this function?  Doesn't look like the contexts are used for anything other than timeout functionality.

Reply:
Yes for now, but the idea I have at the back of my head is to use it with `cancel()` in the future once it's more event-driven.

Context offer the flexibility of timeouts + event driven cancellation, and inherit from each other, so we should really be using them in e2e tests.

Reply:
Further down the line in my branches, this context is actually used to cancel operations, for example if a node crashes unexpectedly.

Reply:
I'd argue that we shouldn't be using context for cancellation.  Dave Cheney wrote about it in his [blog](https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation). 

> The documentation for the context package strongly recommends that context.Context is only for request scoped values:

 With regards to cascading cancellation, you can just have a cancel channel that is closed and can be listened on by multiple goroutines.  

I know using contexts for cancellation is a common pattern.  It's something to keep an eye on for Go 2.  They may remove the cancellation from the context package.

Reply:
Interesting blog post, and I 100% agree context values are the fruit of demons developers :imp: 
It's also sad they didn't figure out how to wait for a goroutine to finish on context cancellation, forcing to still have another signal/error channel to wait.

Now for the closing of a signal channel, there are two limitations I can think of:

- Handling timers to close a channel is painful (always need to check to empty the timer channel as well), and even more when nesting them.
- You cannot close a channel multiple times, so you would have to pretty much re-invent contexts for this (check if the channel is closed etc.)
- The standard library accepts contexts to cancel IO operations, but not close channels and often have no easy `Stop` method. One could use a signal channel to cancel a context down the callers, but it feels wrong to me.